File: backend/src/main/java/org/globe42/domain/Income.java
package org.globe42.domain;

import java.math.BigDecimal;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.ManyToOne;
import javax.persistence.SequenceGenerator;
import javax.validation.constraints.NotNull;

/**
 * An income that a person has, provided by an income source.
 * @author JB Nizet
 */
@Entity
public class Income {
    private static final String INCOME_GENERATOR = "IncomeGenerator";

    @Id
    @SequenceGenerator(name = INCOME_GENERATOR, sequenceName = "INCOME_SEQ", initialValue = 1000, allocationSize = 1)
    @GeneratedValue(generator = INCOME_GENERATOR)
    private Long id;

    /**
     * The person which benefits from this income
     */
    @ManyToOne
    @NotNull
    private Person person;

    /**
     * The source of the income
     */
    @ManyToOne
    @NotNull
    private IncomeSource source;

    /**
     * The monthly amount of the income
     */
    @NotNull
    private BigDecimal monthlyAmount;

    public Income() {
    }

    public Income(Long id) {
        this.id = id;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public Person getPerson() {
        return person;
    }

    public void setPerson(Person person) {
        this.person = person;
    }

    public IncomeSource getSource() {
        return source;
    }

    public void setSource(IncomeSource source) {
        this.source = source;
    }

    public BigDecimal getMonthlyAmount() {
        return monthlyAmount;
    }

    public void setMonthlyAmount(BigDecimal monthlyAmount) {
        this.monthlyAmount = monthlyAmount;
    }
}

File: backend/src/main/kotlin/org/globe42/domain/Income.kt
package org.globe42.domain

import java.math.BigDecimal
import javax.persistence.*
import javax.validation.constraints.NotNull

private const val INCOME_GENERATOR = "IncomeGenerator"

/**
 * An income that a person has, provided by an income source.
 * @author JB Nizet
 */
@Entity
class Income {

    @Id
    @SequenceGenerator(
        name = INCOME_GENERATOR,
        sequenceName = "INCOME_SEQ",
        initialValue = 1000,
        allocationSize = 1
    )
    @GeneratedValue(generator = INCOME_GENERATOR)
    var id: Long? = null

    /**
     * The person which benefits from this income
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @NotNull
    lateinit var person: Person

    /**
     * The source of the income
     */
    @ManyToOne(fetch = FetchType.LAZY)
    @NotNull
    lateinit var source: IncomeSource

    /**
     * The monthly amount of the income
     */
    @NotNull
    lateinit var monthlyAmount: BigDecimal

    constructor()

    constructor(id: Long) {
        this.id = id
    }
}
